src/server/spatial_map.cpp

```diff
@@ -0,0 +1,172 @@
+	if (postIteration)
+		key = SpatialKey(pos.X, pos.Y, pos.Z, true);
+	m_cached.insert({key, id});
+}
+
+// Invalidates upon position update
+void SpatialMap::updatePosition(u16 id, const v3f &oldPos, const v3f &newPos)
+{
+	// Try to leave early if already in the same bucket:
+	auto range = m_cached.equal_range(SpatialKey(newPos));
+	for (auto it = range.first; it != range.second; ++it) {
+		if (it->second == id) {
+			return; // all good, let's get out of here
+		}
+	}
+
+	remove(id, oldPos); // remove from old cache position
+	insert(id, newPos); // reinsert
+}
+
+void SpatialMap::remove(u16 id, const v3f &pos, bool postIteration)
+{
+	if (m_iterators_stopping_insertion_and_deletion) {
+		m_pending_deletes.insert(SpatialKey(pos, id));
+		return;
+	}
+
+	SpatialKey key(pos);
+	if (postIteration)
+		key = SpatialKey(pos.X, pos.Y, pos.Z, true);
+	if(m_cached.find(key) != m_cached.end()) {
+		auto range = m_cached.equal_range(key);
+		for (auto it = range.first; it != range.second; ++it) {
+			if (it->second == id) {
+				m_cached.erase(it);
+				return; // Erase and leave early
+			}
+		}
+	}
+
+	remove(id); // should never be hit
+}
+
+void SpatialMap::remove(u16 id)
+{
+	if(m_iterators_stopping_insertion_and_deletion) {
+		m_pending_deletes.insert(SpatialKey(v3f(), id));
+		return;
+	}
+
+	for (auto it = m_cached.begin(); it != m_cached.end(); ++it) {
+		if (it->second == id) {
+			m_cached.erase(it);
+			return; // Erase and leave early
+		}
+	}
+}
+
+void SpatialMap::removeAll()
+{
+	if(m_iterators_stopping_insertion_and_deletion) {
+		m_remove_all = true;
+	} else {
+		m_cached.clear();
+	}
+}
+
+void SpatialMap::getRelevantObjectIds(const aabb3f &box, const std::function<void(u16 id)> &callback)
+{
+	if (m_cached.empty()) return;
+
+	// When searching, we must round to maximum extent of relevant mapblock indexes.
+	// Since we're using floats, always assume +-1
+	auto low = [](f32 val) -> s16 {
+		s16 _val = static_cast<s16>(val / BS);
+		return (_val >> 4) - 1;
+	};
+	auto high = [](f32 val) -> s16 {
+		s16 _val = static_cast<s16>(val / BS);
+		return (_val >> 4) + 1;
+	};
+
+	v3s16 min(low(box.MinEdge.X), low(box.MinEdge.Y), low(box.MinEdge.Z)),
+		max(high(box.MaxEdge.X), high(box.MaxEdge.Y), high(box.MaxEdge.Z));
+		
+	// We should only iterate using this spatial map when there are at least 1 objects per mapblocks to check.
+	// Otherwise, might as well just iterate.
+
+	v3s16 diff = max - min;
+	uint64_t number_of_mapblocks_to_check = std::abs(diff.X) * std::abs(diff.Y) * std::abs(diff.Z);
+	if(number_of_mapblocks_to_check <= m_cached.size()) { // might be worth it
+		for (s16 x = min.X; x < max.X;x++) {
+			for (s16 y = min.Y; y < max.Y;y++) {
+				for (s16 z = min.Z; z < max.Z;z++) {
+					SpatialKey key(x,y,z, false);
+					if (m_cached.find(key) != m_cached.end()) {
+						m_iterators_stopping_insertion_and_deletion++;
+						auto range = m_cached.equal_range(key);
+						for (auto &it = range.first; it != range.second; ++it) {
```

Comment:
what's the advantage of this callback architecture as opposed to just returning a vector of IDs? Could get rid of the pending lists that way.

Reply:
The vector allocation is costly, that's all. In many places in the current codebase, we do everything in the "filter" callback and just return "False" on every hit, so that there is no vector built, because it's so slow.

Reply:
lambda's are not free either! a C++ object needs to be constructed for those too. And vectors are not *that* expensive.

Reply:
Based on my testing, and apparently other's testing previously, and these benchmarks that are included - the lambda allocation is cheaper than the vector allocation (and reallocations as each is added), and cheaper than just reserving a big vector by default.

Reply:
thank you for verifying it, could I suggest adding a comment to that effect somewhere in the code? Who knows if Minetest will still use Github the next time someone tries to optimize this code.